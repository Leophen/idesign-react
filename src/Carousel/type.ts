export interface CarouselProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 初始轮播项
   * @default 0
   */
  defaultCurrent?: number;
  /**
   * 轮播图的宽度
   * @default 100%
   */
  width?: React.CSSProperties['width'];
  /**
   * 轮播图的高度
   * @default 300
   */
  height?: React.CSSProperties['height'];
  /**
   * 是否自动播放
   * @default false
   */
  autoPlay?: boolean;
  /**
   * 轮播间隔时间
   * @default 3000
   */
  interval?: number;
  /**
   * 是否悬浮时停止轮播
   * @default true
   */
  stopOnHover?: boolean;
  /**
   * 轮播图类型
   * @default default
   */
  type?: 'default' | 'card';
  /**
   * 轮播滑动方向
   * @default horizontal
   */
  direction?: 'horizontal' | 'vertical';
  /**
   * 滑动动画时长
   * @default 300
   */
  duration?: number;
  /**
   * 切换轮播时触发
   */
  onChange?: (current: number) => void;
}

export interface CarouselItemProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 索引值
   * @default 0
   */
  index?: number;
  /**
   * 当前轮播项
   * @default 0
   */
  innerCurrent?: number;
  /**
   * 轮播图类型
   * @default default
   */
  type?: 'default' | 'card';
  /**
   * 是否有动画
   * @default false
   */
  ifAnimation?: boolean;
  /**
   * 轮播项数量
   * @default 1
   */
  childrenLength?: number;
  /**
   * 滑动动画时长
   * @default 300
   */
  duration?: number;
  /**
   * 包裹层宽度
   */
  wrapWidth?: number;
  /**
   * 点击轮播项触发
   */
  clickItem?: (index: number) => void;
}

export interface CarouselNavigationProps {
  /**
   * 轮播项数量
   * @default 0
   */
  itemNum?: number;
  /**
   * 当前轮播项
   * @default 0
   */
  current?: number;
  /**
   * 悬浮导航项时触发
   */
  onEnter?: (current: number) => void;
}
