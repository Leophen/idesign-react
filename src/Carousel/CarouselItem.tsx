import React from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import { CarouselItemProps } from './type';

const CarouselItem: React.FC<CarouselItemProps> = (props) => {
  const {
    children,
    className,
    style,
    index = 0,
    innerCurrent = 0,
    type = 'default',
    ifAnimation = false,
    childrenLength = 1,
    duration = 300,
    wrapWidth = 0,
    clickItem,
    ...restProps
  } = props;

  const disposeIndex = (index: number, innerCurrent: number, childrenLength: number) => {
    if (innerCurrent === 0 && index === childrenLength - 1) {
      return -1;
    }
    if (innerCurrent === childrenLength - 1 && index === 0) {
      return childrenLength;
    }
    if (index < innerCurrent - 1 && innerCurrent - index >= childrenLength / 2) {
      return childrenLength + 1;
    }
    if (index > innerCurrent + 1 && index - innerCurrent >= childrenLength / 2) {
      return -2;
    }
    return index;
  };

  const CARD_SCALE = 210 / 332; // 缩放比例
  const itemWidth = 0.415; // 宽度比例

  const calculateTranslate = (index: number, innerCurrent: number, parentWidth: number, inStage: boolean) => {
    if (inStage) {
      return (parentWidth * ((index - innerCurrent) * (1 - itemWidth * CARD_SCALE) - itemWidth + 1)) / 2;
    }
    if (index < innerCurrent) {
      return (-itemWidth * (1 + CARD_SCALE) * parentWidth) / 2;
    }
    return ((2 + itemWidth * (CARD_SCALE - 1)) * parentWidth) / 2;
  };

  const getZIndex = (isActivity: boolean, inStage: boolean) => {
    if (isActivity) {
      return 2;
    }
    if (inStage) {
      return 1;
    }
    return 0;
  };

  const getItemStyle = () => {
    if (type === 'card') {
      const translateIndex =
        index !== innerCurrent && childrenLength > 2 ? disposeIndex(index, innerCurrent, childrenLength) : index;
      const inStage = Math.round(Math.abs(translateIndex - innerCurrent)) <= 1;
      const translate = calculateTranslate(translateIndex, innerCurrent, wrapWidth, inStage).toFixed(2);
      const isActivity = translateIndex === innerCurrent;
      const zIndex = getZIndex(isActivity, inStage);
      return {
        msTransform: `translateX(${translate}px) scale(${isActivity ? 1 : CARD_SCALE})`,
        WebkitTransform: `translateX(${translate}px) scale(${isActivity ? 1 : CARD_SCALE})`,
        transform: `translateX(${translate}px) scale(${isActivity ? 1 : CARD_SCALE})`,
        transition: `transform ${duration / 1000}s ease`,
        zIndex,
        ...style
      };
    }
    return { ...style };
  };

  return (
    <li
      className={classNames(
        'i-carousel-item',
        className
      )}
      style={getItemStyle()}
      onClick={() => clickItem?.(index)}
      {...restProps}
    >
      {children}
    </li>
  )
}

export default CarouselItem;
