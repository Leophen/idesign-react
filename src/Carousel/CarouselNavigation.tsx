import React from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import { CarouselNavigationProps } from './type';

const CarouselNavigation: React.FC<CarouselNavigationProps> = (props) => {
  const {
    itemNum = 0,
    current = 0,
    onEnter,
  } = props;

  const lists = _.range(0, itemNum)

  const handleEnterItem = (index: number) => {
    onEnter?.(index)
  }

  return (
    <ul className='i-carousel__navigation'>
      {lists.map(item => (
        <li
          className={classNames(
            'i-carousel__navigation-item',
            current - 1 === item && 'i-carousel__navigation-item__active'
          )}
          onMouseEnter={() => handleEnterItem(item)}
          key={item}
        />
      ))}
    </ul>
  )
}

export default CarouselNavigation;
