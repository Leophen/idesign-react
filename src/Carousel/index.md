---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 7
---

# Carousel 轮播图

用于展示多个内容的循环播放轮播组件。

## 基本用法

可通过 `defaultCurrent` 属性控制默认轮播项：

```tsx
import React from 'react';
import { Carousel } from 'idesign-react';

const App = () => {
  const handleChange = (val) => {
    console.log(val)
  }

  return (
    <Carousel defaultCurrent={1} onChange={handleChange}>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>1</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>2</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>3</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>4</div>
      </Carousel.Item>
    </Carousel>
  );
};

export default App;
```

## 垂直布局

可通过 `direction` 属性控制轮播图布局方向：

```tsx
import React from 'react';
import { Carousel } from 'idesign-react';

const App = () => {
  return (
    <Carousel direction="vertical">
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>1</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>2</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>3</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>4</div>
      </Carousel.Item>
    </Carousel>
  );
};

export default App;
```

## 自动轮播

可通过 `autoPlay` 属性控制是否自动轮播，可配合 `interval` 属性控制轮播间隔时间（单位为秒）

```tsx
import React from 'react';
import { Carousel } from 'idesign-react';

const App = () => {
  return (
    <Carousel autoPlay interval={2}>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>1</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>2</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>3</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>4</div>
      </Carousel.Item>
    </Carousel>
  );
};

export default App;
```

## 滑动速度

可通过 `duration` 属性控制滑动速度：

```tsx
import React from 'react';
import { Carousel } from 'idesign-react';

const App = () => {
  return (
    <Carousel duration={100}>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>1</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>2</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>3</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>4</div>
      </Carousel.Item>
    </Carousel>
  );
};

export default App;
```

## 卡片轮播

可通过 `type` 属性指定轮播样式：

```tsx
import React from 'react';
import { Carousel } from 'idesign-react';

const App = () => {
  return (
    <Carousel type='card'>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>1</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>2</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary)', height: '100%' }}>3</div>
      </Carousel.Item>
      <Carousel.Item>
        <div style={{ background: 'var(--i-primary-disabled)', height: '100%' }}>4</div>
      </Carousel.Item>
    </Carousel>
  );
};

export default App;
```

## Carousel API

<API hideTitle />

## CarouselItem API

| 属性      | 说明       | 类型            | 默认值 |
| --------- | ---------- | --------------- | ------ |
| children  | 按钮内容   | `ReactNode`     | `--`   |
| className | 类名       | `string`        | `--`   |
| style     | 自定义样式 | `CSSProperties` | `--`   |
