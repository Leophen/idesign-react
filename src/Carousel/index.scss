.i-carousel {
  position: relative;
  min-height: 60px;
  &.i-carousel__vertical {
    .i-carousel__wrapper {
      flex-direction: column;
    }
    .i-carousel__navigation {
      right: 16px;
      bottom: 50%;
      left: unset;
      flex-direction: column;
      transform: translateY(50%);
      .i-carousel__navigation-item {
        width: 6px;
        height: 26px;
      }
    }
    .i-carousel__arrow-last,
    .i-carousel__arrow-next {
      top: unset;
      right: unset;
      left: 50%;
      transform: translateX(-50%) rotate(90deg);
    }
    .i-carousel__arrow-last {
      top: 16px;
    }
    .i-carousel__arrow-next {
      bottom: 16px;
    }
  }
  &.i-carousel__card {
    .i-carousel__wrapper {
      display: block;
    }
    .i-carousel-item {
      position: absolute;
      top: 0;
      left: 0;
      z-index: 0;
      width: 41.5%;
      height: 100%;
      &.i-carousel-item__active {
        z-index: 2;
      }
    }
  }
}

.i-carousel__content {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
}

.i-carousel__wrapper {
  position: relative;
  display: flex;
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
}

.i-carousel-item {
  display: inline-block;
  flex: 0 0 auto;
  width: 100%;
  height: 100%;
  list-style: none;
}

.i-carousel__navigation {
  position: absolute;
  bottom: 16px;
  left: 50%;
  z-index: 3;
  display: flex;
  gap: 10px;
  align-items: center;
  margin: 0;
  padding: 0;
  transform: translateX(-50%);
}

.i-carousel__navigation-item {
  width: 26px;
  height: 6px;
  list-style: none;
  background: rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  box-shadow: inset 0 1px 6px rgba(0, 0, 0, 0.12);
  cursor: pointer;
  transition: all 0.2s ease-out;
  &.i-carousel__navigation-item__active {
    background: #fff;
  }
}

.i-carousel__arrow-last,
.i-carousel__arrow-next {
  position: absolute;
  top: 50%;
  z-index: 2;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  background: rgba(0, 0, 0, 0.1);
  border-radius: 50%;
  box-shadow: inset 0 1px 6px rgba(0, 0, 0, 0.12);
  transform: translateY(-50%);
  cursor: pointer;
  transition: all 0.2s ease-out;
  &:hover {
    background: rgba(0, 0, 0, 0.2);
  }
}

.i-carousel__arrow-last {
  left: 16px;
}

.i-carousel__arrow-next {
  right: 16px;
}
