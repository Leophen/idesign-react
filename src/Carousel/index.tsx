import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import Icon from '../Icon'
import CarouselItem from './CarouselItem';
import CarouselNavigation from './CarouselNavigation';
import { CarouselProps } from './type';

const Carousel: React.FC<CarouselProps> & { Item: React.ElementType } = (props) => {
  const {
    children,
    className,
    style,
    defaultCurrent = 0,
    width = '100%',
    height = 300,
    autoPlay = false,
    interval = 3,
    stopOnHover = true,
    type = 'default',
    direction = 'horizontal',
    duration = 300,
    onChange,
    ...restProps
  } = props;

  const resetDefaultCurrent = (index: number) => {
    if (type === 'default') {
      if (index >= childrenLength) {
        return 0
      } else if (index <= -1) {
        return childrenLength - 1
      } else {
        return index + 1
      }
    } else {
      return index
    }
  }

  const [innerCurrent, setInnerCurrent] = useState(resetDefaultCurrent(defaultCurrent));

  // 进行子组件筛选，创建子节点列表
  const childrenList = useMemo(
    () =>
      React.Children.toArray(children).filter(
        (child: any) => child.type.displayName === CarouselItem.displayName,
      ),
    [children],
  );
  const childrenLength = childrenList.length;

  const carouselRef = useRef<any>(null);
  const [wrapWidth, setWrapWidth] = useState(0)
  useEffect(() => {
    setWrapWidth(carouselRef.current.getBoundingClientRect().width)
  }, [])

  // 创建渲染用的节点列表
  const carouselItemList = childrenList.map((child: any, index: number) =>
    React.cloneElement(child, {
      key: index,
      index,
      innerCurrent,
      ifAnimation,
      childrenLength,
      wrapWidth,
      type,
      duration,
      clickItem: (index: number) => {
        if (type === 'card') {
          setInnerCurrent(index)
          onChange?.(index)
        }
      },
      ...child.props,
    }),
  );
  // 列表头尾补一项（衔接轮播滑动）
  if (childrenLength > 0 && type === 'default') {
    const firstEle = carouselItemList[0];
    const lastEle = carouselItemList[carouselItemList.length - 1];
    carouselItemList.push(
      React.cloneElement(firstEle, {
        ...firstEle.props,
        key: childrenLength
      }),
    );
    carouselItemList.unshift(
      React.cloneElement(lastEle, {
        ...lastEle.props,
        key: -1
      }),
    );
  }
  const carouselItemLength = carouselItemList.length

  const [ifAnimation, setIfAnimation] = useState(false);
  const animationTimer = useRef<any>(null); // 计时器指针

  // 轮播图通用跳转函数
  const handleTo = useCallback(
    (index: number, handle?: 'last' | 'next') => {
      setIfAnimation(true);
      setInnerCurrent(index);
      if (handle === 'last') {
        if (index > 0) {
          onChange?.(index - 1);
        } else {
          onChange?.(childrenLength - 1);
        }
      } else {
        if (index <= childrenLength) {
          onChange?.(index - 1);
        } else {
          onChange?.(0);
        }
      }
    },
    [childrenLength, onChange],
  );

  // 监听每次轮播结束
  useEffect(() => {
    if (animationTimer.current) {
      clearTimeout(animationTimer.current);
      animationTimer.current = null;
    }
    animationTimer.current = setTimeout(() => {
      setIfAnimation(false);
      if (type !== 'card') {
        if (innerCurrent + 1 >= carouselItemLength) {
          setInnerCurrent(1);
        }
        if (innerCurrent <= 0) {
          setInnerCurrent(carouselItemLength - 2);
        }
      }
    }, duration + 50);
  }, [innerCurrent, carouselItemLength, duration, type]);

  const [ifHoverContent, setIfHoverContent] = useState(false)

  const handleEnterContent = () => {
    stopOnHover && setIfHoverContent(true)
  }

  const handleLeaveContent = () => {
    stopOnHover && setIfHoverContent(false)
  }

  // 自动轮播
  const loopTimer = useRef<any>(null); // 自动轮播指针
  const setLoopTimer = useCallback(() => {
    if (!ifHoverContent && autoPlay && interval > 0) {
      loopTimer.current = setTimeout(
        () => {
          handleTo(innerCurrent + 1);
        },
        innerCurrent === 0 ? interval * 1000 - (duration + 50) : interval * 1000, // 当 index 为 0 时，表示刚从克隆的最后一项跳转过来，且经历了 duration + 50 的间隔时间，减去即可
      );
    }
  }, [autoPlay, ifHoverContent, innerCurrent, duration, interval, handleTo]);
  const clearLoopTimer = useCallback(() => {
    if (loopTimer.current) {
      clearTimeout(loopTimer.current);
      loopTimer.current = null;
    }
  }, []);
  useEffect(() => {
    setLoopTimer();
    return () => clearLoopTimer();
  }, [setLoopTimer])

  const handleClickArrow = (handle: 'last' | 'next') => {
    if (ifAnimation) {
      return false;
    } else {
      if (handle === 'last') {
        if (innerCurrent - 1 < 0) {
          return handleTo(childrenLength - 1, 'last');
        } else {
          return handleTo(innerCurrent - 1, 'last');
        }
      }
      if (handle === 'next') {
        if (type === 'card') {
          return handleTo(innerCurrent + 1 >= carouselItemLength ? 0 : innerCurrent + 1, 'next');
        } else {
          return handleTo(innerCurrent + 1, 'next');
        }
      }
    }
  }

  // Item 包裹层样式
  const getWrapperStyle = () => {
    if (type === 'card') {
      // 卡片样式轮播
      // return {
      //   height: offsetHeight,
      // };
    } else {
      // 默认样式轮播
      if (direction === 'vertical') {
        return {
          transform: `translate3d(0, -${innerCurrent * 100}%, 0px)`,
          transition: ifAnimation ? `transform ${duration / 1000}s ease` : '',
        };
      } else {
        return {
          transform: `translate3d(-${innerCurrent * 100}%, 0px, 0px)`,
          transition: ifAnimation ? `transform ${duration / 1000}s ease` : '',
        };
      }
    }
  };

  return (
    <div
      className={classNames(
        'i-carousel',
        direction === 'vertical' && 'i-carousel__vertical',
        type === 'card' && 'i-carousel__card',
        className
      )}
      style={{ ...style, width, height }}
      ref={carouselRef}
      {...restProps}
    >
      <div
        className="i-carousel__content"
        onMouseEnter={handleEnterContent}
        onMouseLeave={handleLeaveContent}
      >
        <ul
          className='i-carousel__wrapper'
          style={getWrapperStyle()}
        >
          {carouselItemList}
        </ul>
      </div>
      <CarouselNavigation
        itemNum={childrenLength}
        current={type === 'default' ? innerCurrent : innerCurrent + 1}
        onEnter={(current: number) => handleTo(type === 'default' ? current + 1 : current)}
      />
      <div
        className="i-carousel__arrow-last"
        onClick={() => handleClickArrow('last')}
      >
        <Icon color="#fff" name="ArrowLeft" />
      </div>
      <div
        className="i-carousel__arrow-next"
        onClick={() => handleClickArrow('next')}
      >
        <Icon color="#fff" name="ArrowRight" />
      </div>
    </div>
  );
};

Carousel.Item = CarouselItem;

CarouselItem.displayName = 'CarouselItem';
Carousel.displayName = 'Carousel';

export default Carousel;
