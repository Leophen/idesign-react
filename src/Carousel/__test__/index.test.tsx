import { render } from '@testing-library/react';
import Carousel from '../index';

describe('Carousel 组件测试', () => {
  it('create', () => {
    const el = (
      <Carousel defaultCurrent={1}>
        <Carousel.Item>
          item1
        </Carousel.Item>
        <Carousel.Item>
          item2
        </Carousel.Item>
        <Carousel.Item>
          item3
        </Carousel.Item>
      </Carousel>
    );
    const { asFragment } = render(el);
    expect(asFragment()).toMatchSnapshot();
  });
});

