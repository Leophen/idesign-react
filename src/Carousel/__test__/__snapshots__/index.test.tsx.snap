// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-carousel"
    style="width: 100%; height: 300px;"
  >
    <div
      class="i-carousel__content"
    >
      <ul
        class="i-carousel__wrapper"
        style="transform: translate3d(-200%, 0px, 0px);"
      >
        <li
          class="i-carousel-item"
        >
          item3
        </li>
        <li
          class="i-carousel-item"
        >
          item1
        </li>
        <li
          class="i-carousel-item"
        >
          item2
        </li>
        <li
          class="i-carousel-item"
        >
          item3
        </li>
        <li
          class="i-carousel-item"
        >
          item1
        </li>
      </ul>
    </div>
    <ul
      class="i-carousel__navigation"
    >
      <li
        class="i-carousel__navigation-item"
      />
      <li
        class="i-carousel__navigation-item i-carousel__navigation-item__active"
      />
      <li
        class="i-carousel__navigation-item"
      />
    </ul>
    <div
      class="i-carousel__arrow-last"
    >
      <i
        class="i-icon icon-ArrowLeft"
        style="color: rgb(255, 255, 255);"
      />
    </div>
    <div
      class="i-carousel__arrow-next"
    >
      <i
        class="i-icon icon-ArrowRight"
        style="color: rgb(255, 255, 255);"
      />
    </div>
  </div>
</DocumentFragment>
`;
